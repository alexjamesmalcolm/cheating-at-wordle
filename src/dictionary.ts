import dictionaryData from "./dictionary.json";

const notWords = ["MARLY", "DOWVE"];
export const dictionary: string[] = Object.keys(dictionaryData).filter(
  (word) => !notWords.includes(word)
);

export const getWordsOfNLength = (length: number): string[] =>
  dictionary.filter((word) => word.length === length);
