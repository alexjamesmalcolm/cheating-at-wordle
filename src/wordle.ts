import { getWordsOfNLength } from "./dictionary";
import { getRandomElementFromList } from "./getRandomElementFromList";

type Color = "Yellow" | "Green" | "Gray";
export type Result = Color[];

export type Game = (guess: string) => Result;

export const getResultsOfGuess = (correctWord: string, guess: string): Result =>
  guess
    .split("")
    .map((letter) => letter.toUpperCase())
    .map((letter, index): Color => {
      const correctLetter = correctWord.charAt(index);
      if (letter === correctLetter) return "Green";
      if (correctWord.includes(letter)) return "Yellow";
      return "Gray";
    });

export const playWordle = (wordLength: number): Game => {
  const correctWord: string = getRandomElementFromList(
    getWordsOfNLength(wordLength)
  );
  console.log("correct word", correctWord);
  return playFixedWordle(correctWord);
};

export const playFixedWordle = (correctWord: string): Game => {
  const game: Game = (guess) => {
    return getResultsOfGuess(correctWord, guess);
  };
  return game;
};
