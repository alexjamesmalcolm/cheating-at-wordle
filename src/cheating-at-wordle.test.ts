import { describe, it, expect } from "@jest/globals";
import {
  cheatAtWordle,
  doesWordFulfilConstraint,
  getAllWordsUsingConstraints,
} from "./cheating-at-wordle";
import { getWordsOfNLength } from "./dictionary";
// import { getWordsOfNLength } from "./dictionary";
import { playFixedWordle } from "./wordle";

describe("cheatAtWordle", () => {
  it("should guess WANGO", () => {
    const word = "WANGO";
    const game = playFixedWordle(word);
    const result = cheatAtWordle(game, word.length);
    console.log(result);
    expect(result.guesses).toMatchSnapshot();
  });
  it("should guess TAPIR", () => {
    const word = "TAPIR";
    const game = playFixedWordle(word);
    const result = cheatAtWordle(game, word.length);
    console.log(result);
    expect(result.guesses).toMatchSnapshot();
  });
  it("should guess BIRCH", () => {
    const word = "BIRCH";
    const game = playFixedWordle(word);
    const result = cheatAtWordle(game, word.length);
    console.log(result);
    expect(result.guesses).toMatchSnapshot();
  });
  it("should guess LURCH", () => {
    const word = "LURCH";
    const game = playFixedWordle(word);
    const result = cheatAtWordle(game, word.length);
    console.log(result);
    expect(result.guesses).toMatchSnapshot();
  });
  it("should guess PORCH", () => {
    const word = "PORCH";
    const game = playFixedWordle(word);
    const result = cheatAtWordle(game, word.length);
    console.log(result);
    expect(result.guesses).toMatchSnapshot();
  });
  it("should guess POETRY", () => {
    const word = "POETRY";
    const game = playFixedWordle(word);
    const result = cheatAtWordle(game, word.length);
    console.log(result);
    expect(result.guesses).toMatchSnapshot();
  });
});

describe("doesWordFulfilConstraint", () => {
  it("should do something", () => {
    const result = doesWordFulfilConstraint("ZEQUIN", [
      "ZEQUIN",
      ["Gray", "Yellow", "Gray", "Gray", "Yellow", "Gray"],
    ]);
    expect(result).toBe(false);
  });
});

describe("Able to solve every 5 letter word", () => {
  // const fiveLetterWords = getWordsOfNLength(5);
  // const inRange = fiveLetterWords.slice(0, 50);
  // inRange.forEach((word) => {
  //   it(`should solve for ${word}`, () => {
  //     const game = playFixedWordle(word);
  //     const result = cheatAtWordle(game, word.length);
  //     expect(result.won).toBeTruthy();
  //   });
  // });
  it("should solve BERBE", () => {
    const word = "BERBE";
    const game = playFixedWordle(word);
    const result = cheatAtWordle(game, word.length);
    console.log(result.guesses);
    expect(result.won).toBeTruthy();
  });
});

describe("getAllWordsUsingConstraints", () => {
  it("should find good word", () => {
    const words = getAllWordsUsingConstraints(5, [
      ["AROSE", ["Gray", "Gray", "Green", "Yellow", "Gray"]],
      ["SLOTH", ["Green", "Gray", "Green", "Gray", "Gray"]],
      ["SPOON", ["Green", "Yellow", "Green", "Green", "Gray"]],
    ]);
    console.log(words);
    // expect(words).toMatchSnapshot();
  });
  it("should have BLD in the word", () => {
    const words = getWordsOfNLength(5);
    const goodLetters = ["B", "L", "D"];
    const wordsWithCorrectLetters = words.filter((word) =>
      goodLetters.every((goodLetter) => word.includes(goodLetter))
    );
    console.log("words with letters", wordsWithCorrectLetters);
  });
});
