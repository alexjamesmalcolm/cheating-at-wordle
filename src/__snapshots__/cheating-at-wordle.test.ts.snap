// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cheatAtWordle should guess BIRCH 1`] = `
Array [
  "AROSE",
  "RUTIC",
  "CHIRM",
  "BIRCH",
]
`;

exports[`cheatAtWordle should guess LURCH 1`] = `
Array [
  "AROSE",
  "RUTIC",
  "CURLY",
  "LURCH",
]
`;

exports[`cheatAtWordle should guess POETRY 1`] = `
Array [
  "LINEAR",
  "TROUSE",
  "REDTOP",
  "POETRY",
]
`;

exports[`cheatAtWordle should guess PORCH 1`] = `
Array [
  "AROSE",
  "TURIO",
  "CORNY",
  "PORCH",
]
`;

exports[`cheatAtWordle should guess TAPIR 1`] = `
Array [
  "AROSE",
  "HAIRY",
  "CARIB",
  "RAPID",
  "TAPIR",
]
`;

exports[`cheatAtWordle should guess WANGO 1`] = `
Array [
  "AROSE",
  "CANTO",
  "BANJO",
  "WANGO",
]
`;
